### ISSUE TITLE
On the end goal

### ISSUE THREAD

**Issue by oxij** [nix-community/flake.nix#11]
"Flakes" get mentioned a lot lately. I've read the contents of the branches of this repo and discussions in PRs, but I still don't understand what is the issue in the first place and where is the end goal of this?

According to @edolstra at https://github.com/NixOS/nixpkgs/pull/51109#issuecomment-445569542 the idea behind flakes is to replace the `NIX_PATH` with them, according to other quotes general modularization of Nixpkgs. The specification and discussions here talk about versions and version resolution a lot. The specification itself looks like cabal/setup.py files for Nix.

Which takes me to my stupid question. So you have a bunch of packages, each has a bunch of versions and you want to make a consistent set. What is the problem with just using version control for that?

E.g. I store a bunch of git controlled checkout of Nixpkgs/SLNOS/other nix-y things (some of which follow channels, some of which don't) and a bunch of `configuration.nix` files referring to them under a single git controlled meta-repository, which is my `NIX_PATH`. If I need some constraint resolution I just write it in my `nix` files in the meta-repository, most of the time I just update git submodules, commit, bulild and test. Why would I want flakes in such a setup?

In my experience things are pretty tightly coupled everywhere and I usually wish more things were in a single repository, not less. E.g. I wish `nixos-homepage` was a part of `nixpkgs` (maybe not literally, but via some meta-repository where I could make a meta-commit changing both at the same time) because every time I think about editing it I need to change both. 

Similarly, Stackage is essentially a big consistent repository made out of Hackage (that constantly breaks), TeXlive is a similar thing for TeX (made for similar reasons), etc etc. Nixpkgs is a similar thing trying to make a consistent set of everything that can be built with Nix. Nixpkgs itself is already more or less consistent. So why would I want to break Nixpkgs itself into a set of inconsistent repositories?

Sure, with plain version control meta-repository you can't simply transparently use multiple _sets_ of compatible versions, but I don't really see the point of having that functionality most of the time anyway. Think Nixpkgs, most of the time you want a single version of every package, older versions, ideally, are temporary artifacts. Similarly, with version control, if you need a different set of versions, you can just add a new checkout to your meta-repository (or, if its really major, make a new branch in the meta-repository).

Introducing generic constraint solvers just complicates things, IMHO. I'm yet to see a single Cabal, Node or Python package with more than a handful of dependencies that correctly describes its own version requirements. If you really really need some constraint solving, why not make the meta-repository do it like I do in my setup?

/cc @shlevy @zimbatm

P.S. On meta-repository thing: to be even more extremist, let me say I dream of the time when I could have true transparency from Nix to the version control. Think `stdenv.mkDerivation`, I dream to have no `src = fetchurl` lines ever, but, essentially, `src = ./source` where `./source` would be a git submodule (or similar) pointing to the source repository, but so that Nix wouldn't copy it into the store every time on eval. That is to say, I want "version controlled repository" to be a first-class citizen in Nix, and I want to share git objects (or similar) between version control and /nix/store.

**Comment by zimbatm** [comment]
The current implementation of flakes.nix doesn't do much dependency resolution so it's fairly equivalent to your git submodule pinning solution, except that git is not a hard requirement. Maybe somebody wants to use mercurial or another VCS to do software development.

Having everything in a monorepo is certainly advantageous. nixpkgs is benefiting from that a lot I think (and I agree that is would be nice if nixos-homepage was also community-controlled). But there are cases where it's better to use another repo. There are cases, for example with use-like-flags where the complexity explosion wouldn't be worth having inside of nixpkgs. Or some package set like nixpkgs-wayland is only using bleeding-edge packages. ... it's good to be able to have a more distributed development as well.

I don't know if you are using nmattia/niv, it also works great for pinning dependencies. But quickly the issue becomes transitive dependencies. If package A is using nixpkgs-1 and package B is using nixpkgs-2 then it would be great to be able to resolve nixpkgs to a single instance to avoid the long evaluation. That's what flake.nix is solving.

**Comment by oxij** [comment]
> Maybe somebody wants to use mercurial or another VCS to do software development.

Sure, AFAIK most modern VCSes support submodules, mercurial can even make a submodule out of almost anything.

> There are cases, for example with use-like-flags where the complexity explosion wouldn't be worth having inside of nixpkgs. Or some package set like nixpkgs-wayland is only using bleeding-edge packages.

If it is a separate set of packages, then see above (i.e. "why not use git for that?" argument), if it is not, as in use-flags scenario, then I'm even more strongly in favor of the monorepo approach anyway since complicated configurations benefit from a common expressions _more_ than the simple ones. Nobody would want a 100 separate slightly different ffmpeg packages compiled with slightly different options, I hope.

> ... it's good to be able to have a more distributed development as well.

Linux, for instance, does a distributed development in a monorepo. These are not mutually exclusive.

> I don't know if you are using nmattia/niv, it also works great for pinning dependencies. But quickly the issue becomes transitive dependencies. If package A is using nixpkgs-1 and package B is using nixpkgs-2 then it would be great to be able to resolve nixpkgs to a single instance to avoid the long evaluation. That's what flake.nix is solving.

Well, NIX_PATH with `import <package>` (or explicit argument passing) is solving the same problem with pure `git`, which is the point of the OP message.

**Comment by zimbatm** [comment]
Don't get me wrong, I understand your argument; flakes adds a whole new layer of complexity so it's good to question if it brings something that can't be implemented with existing tools.

For SLNOS, do you have nixpkgs as a git submodule? I imagine that if nixpkgs was split into more components you could pull in just the things that you deem worthy of SLNOS

**Comment by oxij** [comment]
> For SLNOS, do you have nixpkgs as a git submodule? I imagine that if nixpkgs was split into more components you could pull in just the things that you deem worthy of SLNOS

No, SLNOS is a set of patchsets on top of nixpkgs (similarly to relationship between tor-browser and firefox, linux and its feature forks, etc), so splitting nixpkgs up in many pieces will only make things more complicated.

**Comment by Mic92** [comment]
It makes a lot sense for single packages that want to provide an installation from source:

While you can install `nix-update` from nixpkgs, you can also get the latest master from `nix run github:Mic92/nix-update`. Also having pinning in NixOS makes a lot of sense. I can build my servers consistently from in my CI: https://github.com/Mic92/dotfiles/blob/master/flake.nix and they will be the same as if I start triggering the evaluation on my laptop. For NixOS I am not aware of any tooling that allows to pin nixpkgs without using hacks within the configuration.nix. The only thing I am worried about is that people are to lazy to update their `flake.nix` and we end up in a similar situation as with all the out-of-date docker images.
